### ISSUE TITLE
Saturation requires unwanted overhead

### ISSUE THREAD

**Issue by rsheeter** [googlefonts/colr-gradients-spec#344]
In discussin w/@behdad today the concern was raised that https://github.com/googlefonts/colr-gradients-spec/blob/main/OFF_AMD2_WD.md#changes-to-off-717-algorithm-for-interpolation-of-instance-values "Saturation arithmetic shall be used: values shall not wrap from maximum to minimum values, or vice versa." is unlikely to be implemented due to overhead. It would be typical to just compute in a relatively wide data type.

Consider removing this language, behavior in the face of going out of bounds is implementation defined.

**Comment by PeterConstable** [comment]
Would it work to change the "shall" wording to "should"? If an implementation uses a wide data type, they'd end up following the recommendation without needing to add any specific validation. And by pointing it out, an implementer might be less likely to use a smaller type that might lead to pitfalls.

**Comment by rsheeter** [comment]
@behdad wdyt? - to me "should not wrap" seems fairly reasonable. Maybe we just delete the saturation part and only say "should not wrap"?

**Comment by behdad** [comment]
I don't like anything prescriptive about this. I think what is needed is just a "note" to say that variations might make the computed value of a fit to go beyond the ranges of the default value.  The PaintRotate can be used as an example.

Let's take a FWORD example. Implementations might choose to use a 32bit integer to represent them, or 16bit, or float. Depending on which one is used, the system will be able to tolerate a wider range of fonts. But I wouldn't expect anyone using 16bit font-space coordinates to implement saturation. If a font hits that limit, then that font is not supported by this rasterizer, end of story. The spec should not be in the business of prescribing what should happen in that case.

**Comment by PeterConstable** [comment]
The _only_ thing the text is currently saying is not to wrap on overflow (if the internal representation is such that there is a chance for overflow). The intent is to improve interoperability of fonts by avoiding a situation in which some implementation wraps and then some fonts try to take advantage of that. It's not clear to me how saying "don't wrap" adds overhead.

**Comment by behdad** [comment]
> The _only_ thing the text is currently saying is not to wrap on overflow

What the actuall eff? If that's what it only says, why is there the word "saturation" in there?

**Comment by PeterConstable** [comment]
>If that's what it only says, why is there the word "saturation" in there?

Your comment seems to indicate that "use saturation" means something significantly different from "don't wrap", but I'm not sure what that is. Please clarify.

**Comment by behdad** [comment]
> Your comment seems to indicate that "use saturation" means something significantly different from "don't wrap", but I'm not sure what that is. Please clarify.

Of course those are different. Here's some code applying variations:
```
  v += delta_contribution;
```

"don't wrap" can for example be interpretted as "use wide-enough data type".

"use saturation" on the other hand, would /require/ (for integer types):
```
  v = SATURATED_ADD(v, delta_contribution);
```

When Rod / I talk about "unwanted overhead", that's what we mean.

I'm going to continue discussion in https://github.com/googlefonts/colr-gradients-spec/pull/350, but the bigger point is: how can you possibly in the name of interoperability specify how to handle overflows when the type and width of the values is left unspecified?!!!  It just doesn't make sense.

**Comment by PeterConstable** [comment]
fixed by #350
